text/0030-standardized-sender-sig.md

```diff
@@ -0,0 +1,137 @@
+
+We have a mechanism for doing this in the memo field, which attaches fixed-size payloads to TxOut's.
+However, this imposes fairly strict size requirements on the data that can be sent.
+
+Outside of the memo field, we have a notion of "receipts" and "confirmation numbers" which a sender
+can use to identify themselves to a recipient. These are hashes of the TxOut shared secret, which
+are similarly deniable, and convince the recipient that whoever sent them this number must have access
+to secrets involved in the construction of the TxOut.
+
+However, these numbers are useless for convincing a third-party who is not a party to the transaction,
+because only the recipient is able to validate these numbers. There is currently no form of receipt that
+can be validated just from the blockchain.
+
+This is desirable by some app project developers, who would like to store encrypted metadata per TxOut,
+but off-chain. For example, they might like to have a database of encrypted metadata, keyed on TxOut's sent
+by users of their app, and be able to ensure that only the true sender of the TxOut can post metadata associated
+to the TxOut, without learning anything about the sender, recipient, amount, etc. of the TxOut.
+
+After this proposal, a straightforward way to do this is that their app can sign the encrypted metadata using
+the tx private key, and their server can check this signature against the public key in the encrypted fog hint.
+We can simply use the Schnorrkel-based signature with Ristretto curve points that we already use in fog public addresses.
+
+This might also be useful because it makes the notion of receipts somewhat more robust -- arbitrary metadata
+can be signed by the sender, and this form of signature will not have the deniability property.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This MCIP will:
+
+* Make it possible for the builder of a transaction output to additionally sign a blob
+    using the tx private key.
+* Make it possible for anyone to validate that signature against the TxOut, without
+    knowing the identity of the sender.
+
+# Reference-level explanation
+[Reference-level-explanation]: #reference-level-explanation
+
+We propose to
+
+* Make internal changes to the transaction builder, following MCIP #18
+  * Instead of being a second independent random private key, the private key used
+    for `mc-crypto-box` encryption will be the same as the tx private key for that
+    `TxOut`.
+  * Also, implement code for a "Janus test" which enables clients to detect the
+    Janus attack, as described in MCIP #18.
+* Standardize a Schnorrkel-based signature using Ristretto curve points, made using the tx private key,
+  and validated using the ephemeral key in the fog hint. This entails:
+  * Make the transaction builder either optionally return the tx private key when `add_output` is called,
```

Comment:
Slight preference for signing at the time `add_output` is called, so that we don't have to build infrastructure around storing the private key for later use. Not sure whether we need to add anything to the MCIP about those tradeoffs, except maybe to consider the challenges of properly storing private key material in infrastructure.

Reply:
The other side of the tradeoff is that if you want to send encrypted data about a previous TXO later, you may end up encrypting data for old TXOs when sending new TXOs, which could get messy to tease apart in the clients, or may not be so bad if we have a reasonable schema for this.

Reply:
So what I had in mind was not that infrastructure stores the key, but that the app can create and submit a transaction, and then sign something perhaps a second or two later that goes to the app server.

It might be that the app needs to hit some endpoint to get some of the data that will be signed. If we allow that they can hold onto the ephemeral private key after the `TransactionBuilder` is gone then they might be able to pipeline their calls effectively. If not then they might be forced to serialize them.

It's just a thought experiment, I don't know if it would matter in a real app.

Reply:
Ok cool - that makes sense to me